README.md
# webmorphR.stim

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

Stimulus sets for use with [webmorphR](https://debruine.github.io/webmorphR).

## Installation

You can install the development version of webmorphR.stim from GitHub with:

``` r
remotes::install_github("debruine/webmorphR.stim")
```

Please cite the individual stimulus sets if you use them in publications:

### neutral

Images are of 102 adult faces 1350x1350 pixels in full colour. Template files mark out 189 coordinates delineating face shape, for use with Psychomorph or WebMorph.org.

Self-reported age, gender and ethnicity are included in `london_info`.

All individuals gave signed consent for their images to be "used in lab-based and web-based studies in their original or altered forms and to illustrate research (e.g., in scientific journals, news media or presentations)." Images were taken in London, UK, in April 2012.
    
<cite>CC-BY: DeBruine, L., & Jones, B. (2017). Face Research Lab London Set (Version 5). figshare. <https://doi.org/10.6084/m9.figshare.5047666.v5></cite>

``` r
load_stim_neutral()
```

![](man/figures/london.jpg)

### smiling

Smiling versions of the 102 individuals in the London set.
    
<cite>CC-BY: DeBruine, L., & Jones, B. (2017). Face Research Lab London Set (Version 5). figshare. <https://doi.org/10.6084/m9.figshare.5047666.v5></cite>

``` r
load_stim_smiling()
```

![](man/figures/smiling.jpg)


### composite

This image set comprises 5 male and 5 female composite images (4 individuals per composite) of adult individuals of African, East Asian, West Asian and White ethnic backgrounds from the London set above, plus a composite image of all 4 ethnic groupings. Images are 1350x1350 pixels in full colour. 

<cite>CC-BY: DeBruine, Lisa (2016): Young adult composite faces. figshare. <https://doi.org/10.6084/m9.figshare.4055130.v1></cite>

``` r
load_stim_composite()
```

![](man/figures/composite.jpg)

### 3dsk

This image set comprises 50 White men and 50 White women ages 18-35. These images are derived from images at [www.3D.sk](https://www.3D.sk), purchased by Lisa DeBruine and Benedict Jones under their [commercial license](https://www.3d.sk/static/commercial-terms), which allows for sharing and publication of derivative images for *educational use*. 

<cite>CC-BY: DeBruine L, Jones B (2020). 3DSK face set with webmorph templates. OSF.
  <https://doi.org/10.17605/OSF.IO/A3947></cite>

``` r
load_stim_3dsk()
```

![](man/figures/3dsk.jpg)

### lisa

CC-0: Four images of the same person with varying face size, position and rotation.

``` r
load_stim_lisa()
```

![](man/figures/lisa.jpg)

### zoom

CC-0: Four images of the same person with varying head-to-camera distance (only very approximately 15, 30, 45 and 60 cm).

``` r
load_stim_zoom()
```

![](man/figures/zoom.jpg)

## canada

This image set contains images of 20 male and 20 female young adult White students from a university in Ontario, Canada. The images have been aligned on pupils and delineated using WebMorph.org. All images are JPGs at 1350x1800 pixels.

Self-reported age and gender are included in `canada_info`.

<cite>CC-BY: DeBruine, L., & Jones, B.. (2017).Young Adult White Faces with Manipulated Versions (Version1). figshare. <https://doi.org/10.6084/m9.figshare.4220517.v1></cite>

``` r
load_stim_canada()
```

![](man/figures/canada.jpg)

## rainbow

This image set contains 18 images from [Pixabay](https://pixabay.com), all using the [Pixabay License](https://pixabay.com/service/license/), so they are free for commercial and noncommercial use without attribution, although you can find available attributions in `rainbow_info`. The images are all 640 pixels wide and varying heights.

``` r
load_stim_rainbow()
```

![](man/figures/rainbow.jpg)



NEWS.md
# webmorphR.stim 0.0.0.9002

* Added a `NEWS.md` file to track changes to the package.
* Added 3D.sk images and data: `load_stim_3dsk()`, `avg_3dsk_info`, `raw_3dsk_info`


tests/testthat.R
library(testthat)
library(webmorphR.stim)

test_check("webmorphR.stim")


tests/testthat/test-data.R
test_that("london_info", {
  x <- london_info

  expect_equal(nrow(x), 102)
  expect_equal(names(x), c("face_id", "face_age", "face_gender", "face_eth"))
})

test_that("canada_info", {
  x <- canada_info

  expect_equal(nrow(x), 40)
  expect_equal(names(x), c("face_id", "face_age", "face_gender"))
})


tests/testthat/test-load_stim.R
# load_stim ----
test_that("load_stim", {
  expect_error(load_stim("no"))

  zoom <- load_stim("zoom")
  expect_equal(length(zoom), 4)
  expect_equal(class(zoom), c("stimlist", "list"))

  five <- load_stim("composite", 1:5)
  expect_equal(length(five), 5)
})

# canada ----
test_that("load_stim_canada", {
  all <- load_stim_canada()
  expect_equal(length(all), 40)
  expect_equal(class(all), c("stimlist", "list"))

  five <- load_stim_canada(1:5)
  expect_equal(length(five), 5)
  expect_equal(names(five), names(all)[1:5])

  male <- load_stim_canada("[0-9]{3}m")
  expect_equal(length(male), 20)
  expect_equal(names(male), names(all)[21:40])
})

# london ----
test_that("load_stim_london", {
  all <- load_stim_london()
  expect_equal(length(all), 102)
  expect_equal(class(all), c("stimlist", "list"))

  five <- load_stim_london(1:5)
  expect_equal(length(five), 5)
  expect_equal(names(five), names(all)[1:5])

  bond <- load_stim_london("007")
  expect_equal(length(bond), 1)
  expect_equal(names(bond), "007_03")
})

# smiling ----
test_that("load_stim_smiling", {
  all <- load_stim_smiling()
  expect_equal(length(all), 102)
  expect_equal(class(all), c("stimlist", "list"))

  five <- load_stim_smiling(1:5)
  expect_equal(length(five), 5)
  expect_equal(names(five), names(all)[1:5])

  bond <- load_stim_smiling("007")
  expect_equal(length(bond), 1)
  expect_equal(names(bond), "007_08")
})

# composite ----
test_that("load_stim_composite", {
  all <- load_stim_composite()
  expect_equal(length(all), 10)
  expect_equal(class(all), c("stimlist", "list"))

  five <- load_stim_composite(1:5)
  expect_equal(length(five), 5)
  expect_equal(names(five), names(all)[1:5])

  white <- load_stim_composite("white")
  expect_equal(length(white), 2)
  expect_equal(names(white), c("f_white", "m_white"))
})

# zoom ----
test_that("load_stim_zoom", {
  all <- load_stim_zoom()
  expect_equal(length(all), 4)
  expect_equal(class(all), c("stimlist", "list"))

  two <- load_stim_zoom(1:2)
  expect_equal(length(two), 2)
  expect_equal(names(two), names(all)[1:2])

  x <- load_stim_zoom("15|60")
  expect_equal(length(x), 2)
  expect_equal(names(x), c("15cm", "60cm"))
})

# lisa ----
test_that("load_stim_lisa", {
  all <- load_stim_lisa()
  expect_equal(length(all), 4)
  expect_equal(class(all), c("stimlist", "list"))

  two <- load_stim_lisa(1:2)
  expect_equal(length(two), 2)
  expect_equal(names(two), names(all)[1:2])

  x <- load_stim_lisa("1|4")
  expect_equal(length(x), 2)
  expect_equal(names(x), c("lisa1", "lisa4"))
})

# rainbow ----
test_that("load_stim_rainbow", {
  all <- load_stim_rainbow()
  expect_equal(length(all), 18)
  expect_equal(class(all), c("stimlist", "list"))

  two <- load_stim_rainbow(1:2)
  expect_equal(length(two), 2)
  expect_equal(names(two), names(all)[1:2])

  x <- load_stim_rainbow("reptile|poppy")
  expect_equal(length(x), 2)
})




R/data.R
#' London Set Information
#'
#' A dataset containing information on the 102 faces in the Face Research Lab London Set. Corresponds to demo sets `load_stim_neutral()` and`load_stim_smiling()`.
#'
#' @format A data frame with 102 rows and 4 variables:
#' \describe{
#'   \item{face_id}{face's ID}
#'   \item{face_gender}{face's gender (female, male)}
#'   \item{face_age}{face's age (19-47 years)}
#'   \item{face_eth}{face's ethnic group (black, east_asian, west_asian, white)}
#' }
#' @source \url{https://figshare.com/articles/dataset/Face_Research_Lab_London_Set/5047666?file=27397184}
#' @examples
#' head(london_info)
"london_info"

#' Canada Set Information
#'
#' A dataset containing information on the 40 faces in the Young Adult White Faces with Manipulated Versions image set. Corresponds to demo set `load_stim_canada()`.
#'
#' @format A data frame with 40 rows and 3 variables:
#' \describe{
#'   \item{face_id}{face's ID}
#'   \item{face_gender}{face's gender (female, male)}
#'   \item{face_age}{face's age (17-35 years)}
#' }
#' @source \url{https://figshare.com/articles/dataset/Young_Adult_White_Faces_with_Manipulated_Versions/4220517?file=7825783}
#' @examples
#' head(canada_info)
"canada_info"

#' Rainbow Set Information
#'
#' A dataset containing information on the 18 images in the rainbow image set `load_stim_rainbow()`. These images are from [Pixabay](https://pixabay.com), all using the [Pixabay License](https://pixabay.com/service/license/), so they are free for commercial and noncommercial use without attribution.
#'
#' @format A data frame with 18 rows and 8 variables:
#' \describe{
#'   \item{photo_name}{the name of the phto inthe simulus set}
#'   \item{photo_id}{the numeric ID from Pixabay}
#'   \item{photographer}{the photographer's name}
#'   \item{user_name}{the photographer's pixabay username}
#'   \item{user_id}{the photographer's pixabay user ID}
#'   \item{colour}{the general colour of the image}
#'   \item{type}{whether the image is an animal, landscape, or flower}
#'   \item{url}{the pixabay URL for the image}
#' }
#' @examples
#' head(rainbow_info)
"rainbow_info"


#' 3DSK Average Ratings
#'
#' A dataset containing information on the 100 images in the 3dsk image set `load_stim_3dsk()`. These images are from [www.3D.sk](https://www.3D.sk), purchased by Lisa DeBruine and Benedict Jones under their [commercial license](https://www.3d.sk/static/commercial-terms), which allows for sharing and publication of derivative images for educational use. Ratings are from 100 male and 100 female raters for each face_gender:trait on a scale from 1-7. See `raw_3dsk_info` for the unaggregated data and further details.
#'
#' @format A data frame with 100 rows and 8 variables:
#' \describe{
#'   \item{face_id}{the name ID from 3d.sk}
#'   \item{face_gender}{face's gender (female, male)}
#'   \item{face_age}{the face's age (in years)}
#'   \item{att}{mean attractiveness rating}
#'   \item{dom}{mean dominance rating}
#'   \item{health}{mean health rating}
#'   \item{sexdim}{mean sexual dimorphism rating; masculinity for male faces and femininity for female faces}
#'   \item{trust}{mean trustworthiness rating}
#' }
#' @source \url{https://doi.org/10.17605/OSF.IO/A3947}
#' @examples
#' head(avg_3dsk_info)
"avg_3dsk_info"


#' 3DSK Raw Ratings
#'
#' A dataset containing information on the 100 images in the 3dsk image set `load_stim_3dsk()`. These images are from [www.3D.sk](https://www.3D.sk), purchased by Lisa DeBruine and Benedict Jones under their [commercial license](https://www.3d.sk/static/commercial-terms), which allows for sharing and publication of derivative images for educational use.
#'
#' @details
#' Face images were rated online at faceresearch.org. Raters were randomly allocated to rate either the 50 male or 50 female faces for one of attractiveness, dominance, health, sexual dimorphism (masculinity/femininity), or trustworthiness using 1 (much less *trait adjective* than average) to 7 (much more *trait adjective* than average). Male faces were rated for masculinity and female faces for femininity.
#'
#' Data are from the first one hundred heterosexual men and one hundred heterosexual women between the ages of 18 and 40 rating each combination of face gender and trait (2000 raters in total). Trial order was fully randomised for each rater and the rating tasks were self paced.
#'
#' @format A data frame with 100000 rows and 8 variables:
#' \describe{
#'   \item{face_id}{the name ID from 3d.sk}
#'   \item{face_gender}{face's gender (female, male)}
#'   \item{face_age}{the face's age (in years)}
#'   \item{rater_id}{the rater's anonymous ID}
#'   \item{rater_gender}{rater's gender (female, male)}
#'   \item{rater_age}{the rater's age (in years)}
#'   \item{trait}{trait being rated (att, dom, health, sexdim, trust)}
#'   \item{rating}{rating on a 1-7 scale}
#' }
#' @source \url{https://doi.org/10.17605/OSF.IO/A3947}
#' @examples
#' head(raw_3dsk_info)
"raw_3dsk_info"


R/load_stim.R
#' Load image set
#'
#' Load image sets as a stimlist, ready for processing with webmorphR functions. See Details below for citation links.
#'
#' @details
#'
#' * neutral/smiling: CC-BY <https://doi.org/10.6084/m9.figshare.5047666.v5>
#' * composite: CC-BY <https://doi.org/10.6084/m9.figshare.4055130.v1>
#' * canada: CC-BY <https://doi.org/10.6084/m9.figshare.4220517.v1>
#' * 3dsk: <https://doi.org/10.17605/OSF.IO/A3947>, via <https://www.3d.sk> for educational use only
#' * lisa/zoom: CC-0 images of the package creator
#' * rainbow: see `rainbow_info` for individual image credits
#'
#'
#' @param set name of the image set
#' @param pattern Vector of patterns to use to search for files, or a vector of image indices
#'
#' @return a list with class stimlist
#' @export
#'
#' @examples
#' # loads the first 5 images in the set
#' stim <- load_stim("rainbow", 1:5)
#'
#' # loads all images with 3 digits and m
#' canada_male <- load_stim("canada", "[0-9]{3}m")
load_stim <- function(set = c("neutral_front", "smiling_front", "3dsk",
                              "composite", "canada", "lisa", "zoom", "rainbow"),
                      pattern = NULL) {
  set <- match.arg(set)
  path <- system.file(set, package = "webmorphR.stim")
  webmorphR::read_stim(path, pattern)
}

#' @export
#' @rdname load_stim
load_stim_canada <- function(pattern = NULL) {
  load_stim("canada", pattern)
}

#' @export
#' @rdname load_stim
load_stim_london <- function(pattern = NULL) {
  load_stim("neutral_front", pattern)
}

#' @export
#' @rdname load_stim
load_stim_neutral <- function(pattern = NULL) {
  load_stim("neutral_front", pattern)
}

#' @export
#' @rdname load_stim
load_stim_zoom <- function(pattern = NULL) {
  load_stim("zoom", pattern)
}

#' @export
#' @rdname load_stim
load_stim_smiling <- function(pattern = NULL) {
  load_stim("smiling_front", pattern)
}

#' @export
#' @rdname load_stim
load_stim_composite <- function(pattern = NULL) {
  load_stim("composite", pattern)
}

#' @export
#' @rdname load_stim
load_stim_lisa <- function(pattern = NULL) {
  load_stim("lisa", pattern)
}

#' @export
#' @rdname load_stim
load_stim_rainbow <- function(pattern = NULL) {
  load_stim("rainbow", pattern)
}

#' @export
#' @rdname load_stim
load_stim_3dsk <- function(pattern = NULL) {
  load_stim("3dsk", pattern)
}



data-raw/3dsk.R
library(tidyverse)
# https://osf.io/a3947/

men <- list.files("data-raw/3dsk", "_men_", full.names = TRUE) |>
  readr::read_csv(id = "path") |>
  mutate(path = gsub('data-raw/3dsk/3dsk_men_', '', path, fixed = TRUE),
         path = gsub('\\.csv', '', path),
         face_sex = "male",
         rater_id = faux::make_id(1:1000, prefix = "R")) |>
  pivot_longer(cols = alex:vlastimil,
               names_to = "face_id",
               values_to = "rating") |>
  select(face_id, face_gender = face_sex, rater_id, rater_gender = sex, rater_age = age, trait = path, rating) |>
  mutate(trait = ifelse(trait == "masc", "sexdim", trait))

women <- list.files("data-raw/3dsk", "_women_", full.names = TRUE) |>
  readr::read_csv(id = "path") |>
  mutate(path = gsub('data-raw/3dsk/3dsk_women_', '', path, fixed = TRUE),
         path = gsub('\\.csv', '', path),
         face_sex = "female",
         rater_id = faux::make_id(1001:2000, prefix = "R")) |>
  pivot_longer(cols = agata:zora,
               names_to = "face_id",
               values_to = "rating") |>
  select(face_id, face_gender = face_sex, rater_id, rater_gender = sex, rater_age = age, trait = path, rating) |>
  mutate(trait = ifelse(trait == "fem", "sexdim", trait))

age_data <- read_csv("data-raw/3dsk/age_data.csv") |>
  select(face_id = face, face_age = age)


raw_3dsk_info <- bind_rows(men, women) |>
  left_join(age_data, by = 'face_id') |>
  select(face_id, face_gender, face_age,
         rater_id, rater_gender, rater_age,
         trait, rating)
readr::write_csv(raw_3dsk_info, "data-raw/raw_3dsk_info.csv")
usethis::use_data(raw_3dsk_info, overwrite = TRUE)

avg_3dsk_info <- raw_3dsk_info |>
  group_by(face_id, face_gender, face_age, trait) |>
  summarise(rating = mean(rating, na.rm = TRUE),
            .groups = 'drop') |>
  pivot_wider(names_from = trait,
              values_from = rating) |>
  arrange(face_gender, face_id)

readr::write_csv(avg_3dsk_info, "data-raw/avg_3dsk_info")
usethis::use_data(avg_3dsk_info, overwrite = TRUE)



stuff/figs.R
library(webmorphR)

## london composite
read_stim("inst/london/")  |>
  setnames(pattern = "_03", replacement = "") |>
  mlabel(gravity = "northwest",
         location = "+20+10",
         size = 200,
         strokecolor = "white") |>
  plot(padding = 5, maxwidth = 1000, nrow = 6) |>
  write_stim("man/figures/", "london", "jpg")

## smiling composite
read_stim("inst/smiling/")  |>
  setnames(pattern = "_08", replacement = "") |>
  mlabel(gravity = "northwest",
         location = "+20+10",
         size = 200,
         strokecolor = "white") |>
  plot(padding = 5, maxwidth = 1000, nrow = 6) |>
  write_stim("man/figures/", "smiling", "jpg",
             overwrite = TRUE)

## canada composite
read_stim("inst/canada/")  |>
  mlabel(gravity = "northwest",
         location = "+20+10",
         strokecolor = "white") |>
  plot(padding = 5, maxwidth = 1000, nrow = 4) |>
  write_stim("man/figures/", "canada", "jpg")

## composite composite
read_stim("inst/composite/")  |>
  mlabel(gravity = "northwest",
         location = "+20+10",
         size = 150,
         strokecolor = "white") |>
  plot(padding = 5, maxwidth = 1000, nrow = 2) |>
  write_stim("man/figures/", "composite", "jpg",
             overwrite = TRUE)

## zoom composite
read_stim("inst/zoom/")  |>
  mlabel(gravity = "northwest",
         location = "+20+10",
         size = 150,
         color = "white",
         strokecolor = "black") |>
  plot(padding = 5, maxwidth = 1000, nrow = 1) |>
  write_stim("man/figures/", "zoom", "jpg",
             overwrite = TRUE)

## lisa composite
read_stim("inst/lisa/")  |>
  mlabel(gravity = "southwest",
         location = "+20+10",
         size = 80,
         color = "white",
         strokecolor = "black") |>
  plot(padding = 5, maxwidth = 1000, nrow = 1) |>
  write_stim("man/figures/", "lisa", "jpg",
             overwrite = TRUE)

## rainbow composite
order <- rainbow_info |>
  dplyr::arrange(type, colour) |>
  dplyr::pull(photo_name)

read_stim("inst/rainbow/")[order]  |>
  pad(50, 0, 0, 0) |>
  mlabel(gravity = "north",
         location = "+0+5",
         size = 40,
         color = "black") |>
  plot(padding = 5, maxwidth = 1000, nrow = 3) |>
  write_stim("man/figures/", "rainbow", "jpg",
             overwrite = TRUE)


## 3dsk
s <- load_stim_3dsk()
s |> pad(1, fill = 'black') |>
  pad(200, 0, 0, 0) |>
  label() |>
  plot(nrow = 5) |>
  write_stim("man/figures/", "3dsk", "jpg",
             overwrite = TRUE)

# s2 <- s |> rename_stim(pattern = "(fe)?male/", replacement = "")
# write_stim(s2[1:50], "~/Desktop/3dsk/female", format = "jpg", quality = 90)
# write_stim(s2[51:100], "~/Desktop/3dsk/male", format = "jpg", quality = 90)
